src/api/udf/Histogram.h

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright 2017-present Shawn Cao
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <atomic>
+#include <glog/logging.h>
+#include <folly/stats/Histogram.h>
+
+#include "surface/eval/UDF.h"
+
+namespace nebula {
+namespace api {
+namespace udf {
+
+template <nebula::type::Kind IK,
+          typename Traits = nebula::surface::eval::UdfTraits<nebula::surface::eval::UDFType::HIST, IK>,
+          typename BaseType = nebula::surface::eval::UDAF<Traits::Type, IK>>
+class Hist : public BaseType {
+  // Set default number of buckets to 20
+  static constexpr size_t BUCKET_NUM = 20;
+
+public:
+  using InputType = typename BaseType::InputType;
+  using NativeType = typename BaseType::NativeType;
+  using BaseAggregator = typename BaseType::BaseAggregator;
+
+public:
+  class Aggregator : public BaseAggregator {
+  static constexpr auto SIZE_SIZE = sizeof(size_t);
+
+  public:
+    explicit Aggregator(InputType min, InputType max, size_t bucketNum = BUCKET_NUM)
+      : min_{ min },
+      max_{ max },
+      bucketNum_{ bucketNum },
+      bucketSize_{ static_cast<InputType>((max_ - min_) / bucketNum) },
+      histogram_{ bucketSize_, min_, max_ } {
+    }
+    virtual ~Aggregator() = default;
+
+    inline virtual void merge(InputType v) override {
+      histogram_.addValue(v);
+    }
+
+    inline virtual void mix(const nebula::surface::eval::Sketch& another) override {
+      auto right = static_cast<const Aggregator&>(another);
+      histogram_.merge(right.histogram_);
+    }
+
+    inline virtual NativeType finalize() override {
+      std::ostringstream tsv;
+      histogram_.toTSV(tsv);
+      return static_cast<NativeType>(tsv.str());
```

Comment:
should we consider use JSON format (JSON array) in the final result which is common to use in clients?

Reply:
and also client won't rely on external format defined by folly which may change in the future.

Reply:
thanks! saved as json by using rapidJson